### PR TITLE
Issue/1528: Define an abstract list table base class, used to extend all list tables. #1528

### PR DESCRIPTION
**Pull request by ramiabraham**
- Defines an abstract list table base class, used to extend all list tables.
- Require base class in main plugin loader.
- See: `admin/class-affwp-list-table.php`

#1528

### CODE REVIEW

File: includes/admin/class-affwp-list-table.php [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+	 * @access public
+	 * @since 1.9
+	 *
+	 * @param string $text Label for the search box
+	 * @param string $input_id ID of the search box
+	 *
+	 * @return void
+	 */
+	public function search_box( $text, $input_id ) {
+		if ( empty( $_REQUEST['s'] ) && !$this->has_items() )
```
Comment (pippinsplugins) [inline includes/admin/class-affwp-list-table.php]: Let's please go ahead and add brackets around this.

File: includes/admin/class-affwp-list-table.php [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+			return;
+
+		$input_id = $input_id . '-search-input';
+
+		if ( ! empty( $_REQUEST['orderby'] ) )
```
Comment (pippinsplugins) [inline includes/admin/class-affwp-list-table.php]: Also brackets here please

File: includes/admin/class-affwp-list-table.php [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+	}
+
+	/**
+	 * Retrieve the view types
+	 *
+	 * @access public
+	 * @since 1.9
+	 * @return array $views All the views available for this class.
+	 */
+	public function get_views() {
```
Comment (pippinsplugins) [inline includes/admin/class-affwp-list-table.php]: Could we expand this to make registering views as simple as setting up a single array? That way child classes do not need to perform logic to set the current view.

File: includes/admin/class-affwp-list-table.php [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+	}
+
+	/**
+	 * Retrieve the table columns
+	 *
+	 * @access public
+	 * @since  1.9
+	 * @return array $columns Array of all the list table columns.
+	 */
+	public function get_columns() {
```
Comment (pippinsplugins) [inline includes/admin/class-affwp-list-table.php]: Let's break this into two methods, one that is private (or at least marked as such) and contains the `'affwp_list_table_' . $this->singular . '_table_columns'` filter and then one that is public that does nothing more than return the array of column names / keys.

File: includes/admin/class-affwp-list-table.php [hunk 5]
```diff
@@ -0,0 +1,315 @@
...
+	}
+
+	/**
+	 * Retrieve the table's sortable columns.
+	 *
+	 * @access public
+	 * @since  1.9
+	 * @return array Array of all the sortable columns.
+	 */
+	public function get_sortable_columns() {
```
Comment (pippinsplugins) [inline includes/admin/class-affwp-list-table.php]: Same thing has as mentioned above for get_columns()

File: includes/admin/class-affwp-list-table.php [hunk 6]
```diff
@@ -0,0 +1,315 @@
...
+	}
+
+	/**
+	 * Retrieve the bulk actions
+	 *
+	 * @access public
+	 * @since  1.9
+	 * @return array $actions Array of the bulk actions available for this list table.
+	 */
+	public function get_bulk_actions() {
```
Comment (pippinsplugins) [inline includes/admin/class-affwp-list-table.php]: Same thing here as mentioned above for get_columns()

File: includes/admin/class-affwp-list-table.php (hunk 1 above)
Comment (ramiabraham) [inline includes/admin/class-affwp-list-table.php]: (Updated)

File: includes/admin/class-affwp-list-table.php (hunk 2 above)
Comment (ramiabraham) [inline includes/admin/class-affwp-list-table.php]: (Updated)

File: includes/admin/class-affwp-list-table.php (hunk 3 above)
Comment (ramiabraham) [inline includes/admin/class-affwp-list-table.php]: Added in https://github.com/AffiliateWP/AffiliateWP/pull/1533/commits/469fca21c2ae32b9782178d33ca69a4c03ab669c

File: includes/admin/class-affwp-list-table.php (hunk 5 above)
Comment (ramiabraham) [inline includes/admin/class-affwp-list-table.php]: Added in https://github.com/AffiliateWP/AffiliateWP/pull/1533/commits/c2642f192ff3ec2cf6d050f8137d3ee5411e9a1f

File: includes/admin/class-affwp-list-table.php (hunk 4 above)
Comment (ramiabraham) [inline includes/admin/class-affwp-list-table.php]: Added in https://github.com/AffiliateWP/AffiliateWP/pull/1533/commits/c2642f192ff3ec2cf6d050f8137d3ee5411e9a1f

File: includes/admin/class-affwp-list-table.php (hunk 6 above)
Comment (ramiabraham) [inline includes/admin/class-affwp-list-table.php]: Added in https://github.com/AffiliateWP/AffiliateWP/pull/1533/commits/c2642f192ff3ec2cf6d050f8137d3ee5411e9a1f

**Comment by ramiabraham** [comment]
As of https://github.com/AffiliateWP/AffiliateWP/pull/1533/commits/5a227c1a2628fcd23ab519642c6564a5451cffab:

- Above adjustments added
- Simplified `get_views()` method
- Introduced private methods as needed.

**Pull request closed without merging**